### PR TITLE
chore(TS): Group

### PR DESCRIPTION
**Pull request by ShaMan123**
## Description

Migrate Group & ActiveSelection

## Changes

- removed redundant deep clone of option in `fromObject` 4e8db1adf
- removed `Rect` `TObject` stub types

## Gist

Will have conflicts with #8431 - should resolve with `union` strategy, sort of. I can do it

### CODE REVIEW

**Comment by ShaMan123** [comment]
@asturur do you want me to extract the layout mechanism to a standalone class in a dedicated PR?
Now, after working on controls, I have a better idea using  originX/Y of group to position objects in layout. It should simply logic a great deal, but might add additional iterations over objects.

File: src/mixins/collection.mixin.ts [hunk 1]
```diff
@@ -133,7 +135,9 @@ export function createCollectionMixin(Klass: { new (...args: any[]): any }) {
         return true;
       } else if (deep) {
         return this._objects.some(
-          (obj) => obj instanceof Collection && obj.contains(object, true)
+          (obj) =>
+            obj instanceof Collection &&
+            (obj as unknown as Collection).contains(object, true)
```
Comment (ShaMan123) [inline src/mixins/collection.mixin.ts:140]: TS doesn't manage to cope with this line

File: src/shapes/active_selection.class.ts [hunk 2]
```diff
@@ -1,191 +1,171 @@
...
+  /**
+   * Renders controls and borders for the object
+   * @param {CanvasRenderingContext2D} ctx Context to render on
+   * @param {Object} [styleOverride] properties to override the object style
+   * @param {Object} [childrenOverride] properties to override the children overrides
+   */
+  _renderControls(
+    ctx: CanvasRenderingContext2D,
+    styleOverride?: ControlRenderingStyleOverride,
+    childrenOverride?: ControlRenderingStyleOverride
```
Comment (ShaMan123) [inline src/shapes/active_selection.class.ts:130]: guessed these type

File: src/shapes/group.class.ts [hunk 3]
```diff
@@ -1,1053 +1,1058 @@
...
-        this._applyLayoutStrategy(
-          Object.assign({}, opt, {
-            type: 'object_modified',
-          })
-        );
+  /**
+   * keeps track of the selected objects
+   * @private
+   */
+  __objectSelectionMonitor(selected: boolean, opt) {
```
Comment (ShaMan123) [inline src/shapes/group.class.ts:307]: this can be typed when #8431 is merged

**Review by ShaMan123 — COMMENTED** [review]
Ready to extract layout mechanism or to merge

**Comment by ShaMan123** [comment]
This PR finalizes shapes!

File: src/brushes/pattern_brush.class.ts [hunk 4]
```diff
@@ -73,7 +69,7 @@ export class PatternBrush extends PencilBrush {
       topLeft = path._getLeftTopCoords().scalarAdd(path.strokeWidth / 2);
 
     path.stroke = new Pattern({
-      source: this.source || this.getPatternSrcFunction(),
+      source: this.source || eval(this.getPatternSrcFunction()),
```
Comment (ShaMan123) [inline src/brushes/pattern_brush.class.ts:72]: @asturur what if going on here??
I remember seeing bug reports on this.
I surfaced now that TS checks it.
I suggest refactoring `getPatternSrcFunction` signature to `getPatternSrcFunction(color)` and then we can drop this weird thing replacing the text in the function and evaling it

Comment (ShaMan123) [reply src/brushes/pattern_brush.class.ts:72]: I am going to checkout this change in favor of #8468

File: src/shapes/active_selection.class.ts (hunk 2 above)
Comment (asturur) [reply src/shapes/active_selection.class.ts:131]: seems correct

**Comment by asturur** [comment]
> @asturur do you want me to extract the layout mechanism to a standalone class in a dedicated PR? Now, after working on controls, I have a better idea using originX/Y in relation to group to position objects in layout. It should simplify logic a great deal, but might add additional iterations over objects.

I would like to do that, because that is the only way for me to settle on what to do with those layout strategies.

File: src/brushes/pattern_brush.class.ts (hunk 4 above)
Comment (asturur) [reply src/brushes/pattern_brush.class.ts]: we need to delete this getPatternSrcFunction. We shouldn't support serialiazing code.
It was an old thing i thought it was gone.

File: src/shapes/active_selection.class.ts [hunk 5]
```diff
@@ -1,191 +1,174 @@
...
-        }
-      },
+  /**
+   * @private
+   * @param {'added'|'removed'} type
+   * @param {FabricObject[]} targets
+   */
+  _onAfterObjectsChange(type: 'added' | 'removed', targets: FabricObject[]) {
+    // @TODO figure out this change. This part wasn't here before migration.
+    super._onAfterObjectsChange(type, targets);
```
Comment (asturur) [inline src/shapes/active_selection.class.ts:65]: I m merging now, but i wonder why we added this line here.
This adds an extra layout call that we weren't doing before?

**Review by asturur — APPROVED** [review]

File: src/shapes/active_selection.class.ts (hunk 5 above)
Comment (ShaMan123) [reply src/shapes/active_selection.class.ts:65]: Yes, correct.
It was a bug that I fixed in a pending PR so I decided to put it here

**Pull request merged**